[asl reference] LR1 parsing exposition (#1778)
herd/herdtools7: asllib/doc/Expressions.tex

Reviewer: Also should be updated to match the numbering given in §6.6. _Priority and Associativity_

Reviewer: ```suggestion
  \vmathematicallyassociativeoperators \eqdef \left\{
```

%
Note the distinction between parsing associativity
(as given in \secref{PriorityAndAssociativity})
and mathematical associativity (the operators listed above).

Intuitively, the above rules ensure that an arithmetic expression in ASL is
human-readable: the expression follows normal operator priority rules
of exponentiation before multiply/divide before add/subtract. However:
\begin{itemize}
  \item Potentially unclear examples such as \texttt{3 - 2 - 1} cause static
        errors unless parenthesised to clarify between \texttt{(3 - 2) - 1 = 0}
        and \texttt{3 - (2 - 1) = 2}.
  \item Unnecessary parentheses are minimised in examples such as
        \texttt{1 + 2 + 3}, as both bracketings \texttt{(1 + 2) + 3} and
        \texttt{1 + (2 + 3)} can be considered identical to a reader.
\end{itemize}

See \ASTRuleRef{EBinop} for the formal details.

For convenience, we duplicate here the parsing priority and associativity table
from \secref{PriorityAndAssociativity}, keeping only those rows that contain
binary operators.

\begin{center}
\begin{tabular}{lll}
\hline
\textbf{Priority} & \textbf{Class} & \textbf{Operators}\\
\hline

> 7 & \textbf{Unary}                & \Tminus, \Tnot, \Tbnot\\

Reviewer: I realise above we say "only those rows that contain binary operators", but this line is unary - let's remove this row, unless it's needed?

> $\STRCONCAT$.

Reviewer: I think this needs deleting too

The specification in \listingref{BinaryExpressions} shows examples of binary expressions
and the effect of operator priority.

> \ASLListing{Binary expressions and operator priority}{BinaryExpressions}{\definitiontests/Guide.OperatorPriority.asl}

Reviewer: I think the example needs updating too, at least the following:
- throughout: `precedence` --> `priority`
- clarify that `AND` is parsed as left-associated, but to the reader it doesn't matter

\ASTRuleDef{EBinop}
The following rule constructs a binary expression AST from a parse tree,
which possibly contains parentheses,
ensuring that \RequirementRef{BinaryOperatorPriorities} holds.

The specification in \listingref{BinaryExpressions}
shows examples where \RequirementRef{BinaryOperatorPriorities} holds.
%
The specifications in
\listingref{EBinop-bad1},
\listingref{EBinop-bad2},
\listingref{EBinop-bad3},
\listingref{EBinop-bad4}, and
\listingref{EBinop-bad5}
show examples that violate \RequirementRef{BinaryOperatorPriorities},
which results in a \builderrorterm.
\ASLListing{Violation of operator priority}{EBinop-bad1}{\syntaxtests/ASTRule.EBinop.bad1.asl}
\ASLListing{Violation of operator priority}{EBinop-bad2}{\syntaxtests/ASTRule.EBinop.bad2.asl}
\ASLListing{Violation of operator priority}{EBinop-bad3}{\syntaxtests/ASTRule.EBinop.bad3.asl}
\ASLListing{Violation of operator priority}{EBinop-bad4}{\syntaxtests/ASTRule.EBinop.bad4.asl}

> \ASLListing{Violation of operator priority}{EBinop-bad5}{\syntaxtests/ASTRule.EBinop.bad5.asl}

Reviewer: Similarly, need to update the code and comments in all these examples (including referring to "mathematical associativity"

}
\end{mathpar}

\ASTRuleDef{CheckNotSamePriority}
\hypertarget{build-binopprec}{}
We define the helper function
\[

>   \binopprec(\overname{\binop}{\op}) \aslto \N

Reviewer: ```suggestion
  \binoppri(\overname{\binop}{\op}) \aslto \N
```
(or similar)

> \checknotsameprec(\overname{\binop}{\op} \aslsep \overname{\expr}{\ve})

Reviewer: ```suggestion
\checknotsamepri(\overname{\binop}{\op} \aslsep \overname{\expr}{\ve})
```
(similarly throughout)

\aslto \{\True\} \cup \overname{\TBuildError}{\BuildErrorConfig}
\]

> checks whether the expression AST node $\ve$ is a binary operator, either with a different operator to $\op$ but the same \emph{priority}, or the same non-associative operator $\op$.

Reviewer: ```suggestion
checks whether the expression AST node $\ve$ is a binary operator, either with a different operator to $\op$ but the same \emph{priority}, or the same non-mathematically-associative operator $\op$.
```

is used, whereas \texttt{a - b - c} and \texttt{a + b - c} are considered illegal.

> For the first, this is because $\SUB$ is not associative.

Reviewer: ```suggestion
For the first, this is because $\SUB$ is not mathematically associative.
```

  \begin{itemize}
    \item $\ve$ is a binary operation expression for the operator $\opp$;
    \item $\opp$ is the same operator as $\op$;

>     \item checking whether $\op$ is an associative operator yields $\True$\ProseTerminateAs{\BinopPriority}.

Reviewer: ```suggestion
    \item checking whether $\op$ is a mathematically associative operator yields $\True$\ProseTerminateAs{\BinopPriority}.
```
